private-identity-audit/ffi/src/lib.rs

```diff
@@ -0,0 +1,466 @@
+//! unique identity membership proofs and verifying them as part of the
+//! PIAL project.
+
+#![feature(vec_into_raw_parts)]
+
+extern crate libc;
+use codec::{Decode, Encode};
+use cryptography_core::{
+    cdd_claim::{CddClaimData, CddId},
+    curve25519_dalek::scalar::Scalar,
+};
+use libc::size_t;
+use private_identity_audit::{
+    uuid_to_scalar, CommittedSetGenerator, CommittedUids, PrivateUids, ProofGenerator,
+    ProofVerifier, Prover, Verifier, VerifierSecrets, VerifierSetGenerator, ZKPFinalResponse,
+    ZKPInitialmessage,
+};
+use rand::{rngs::StdRng, SeedableRng};
+use std::{ptr::null_mut, slice};
+use uuid::{Builder, Variant, Version};
+
+/// Used to pass a Vec of u8 between Rust and FFI users.
+/// The code assumes that the pointer `ptr` is the start a "continous" block of memory of size `n`.
+#[repr(C)]
+#[derive(Debug, Clone)]
+pub struct VecEncoding {
+    ptr: *mut u8,
+    n: usize,
+}
+
+impl VecEncoding {
+    pub fn new(vec: Vec<u8>) -> Self {
+        let (ptr, len, _cap) = vec.into_raw_parts();
+        Self { ptr, n: len }
+    }
+    /// Converts the pointer received from FFI to a Vector of u8.
+    ///
+    /// # Safety
+    /// Refer to the safety notes of `Vec::from_raw_parts`.
+    unsafe fn to_vec(&self) -> Vec<u8> {
+        Vec::from_raw_parts(self.ptr, self.n, self.n)
+    }
+}
+
+/// Used to pass a 2D Vec of u8 between Rust and FFI users.
+/// The code assumes that the pointer `ptr` is the start a "continous" block of memory of size `rows*cols`.
+#[repr(C)]
+#[derive(Debug)]
+pub struct MatrixEncoding {
+    ptr: *mut u8,
+    rows: usize,
+    cols: usize,
+}
+
+impl MatrixEncoding {
+    /// The code assumes that all the elements of `vec` encode to the same length.
+    pub fn new<T: Encode>(vec: Vec<T>) -> Option<Self> {
+        let vec = vec
+            .iter()
+            .map(|item| item.encode())
+            .collect::<Vec<Vec<u8>>>();
+        let rows = vec.len();
+        if let Some(cols) = vec.first() {
+            let cols = cols.len();
+            for r in vec.iter() {
+                if r.len() != cols {
+                    println!("Mismatch on encoded data size. Expected all items to encode to {} bytes, found an element that encodes to {} bytes.", cols, r.len());
+                    return None;
+                }
+            }
+
+            let mut flattened_vec: Vec<u8> = Vec::with_capacity(rows * cols);
+            for row in vec.iter() {
+                flattened_vec.extend(row);
+            }
+            assert_eq!(flattened_vec.capacity(), rows * cols);
+            let (ptr, _len, _cap) = flattened_vec.into_raw_parts();
+            Some(Self { ptr, rows, cols })
+        } else {
+            println!("The input `vec` should have at least two elements");
+            None
+        }
+    }
+
+    fn as_slice(&self) -> &[u8] {
+        // SAFETY: Refer to the safety notes of `slice::from_raw_parts_mut`.
```

Comment:
```suggestion
        // SAFETY: If `self` came from Rust, we know we have a valid `self` created in `fn new`,
        // so `ptr` points to a valid `&amp;[u8]` with exactly `rows * cols` elements.
        // Otherwise, if `self` came from outside Rust, e.g., from C, it too has promised somewhere
        // that `self` was created via `new` or equivalent logic.
        //
        // Beyond this, refer to the safety notes of `std::slice::from_raw_parts`.
```